### PR TITLE
Fix ';;' formatting between  doc-comments and toplevel directives

### LINKED ISSUE #2407: Bug: bad interaction between doc comments and toplevel directives
**Describe the bug**
```
$ cat b.mlt
(** Header *)

#use "something"

let two = 2

$ ocamlformat b.mlt  --enable-outside-detected-project --no-version-check --ignore-invalid-option  --debug
[DEBUG] Box structure: /tmp/b_boxes.html
ocamlformat: Cannot process "b.mlt".
  Please report this bug at https://github.com/ocaml-ppx/ocamlformat/issues.
  BUG: generating invalid ocaml syntax.
  output file: /tmp/b.invalid-ast.mlt
  
$ cat /tmp/b.invalid-ast.mlt 
(** Header *);;

#use "something"

let two = 2
```

### PR DESCRIPTION
**Pull request by gpetiot**
Fix #2407

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -31,6 +31,7 @@ Tags:
 - Fix dropped attributes on a begin-end in a match case (#2421, @Julow)
 - Fix non-stabilizing comments before a functor type argument (#2420, @Julow)
 - Fix crash caused by module types with nested `with module` (#2419, @Julow)
+- Fix ';;' formatting between  doc-comments and toplevel directives (#2432, @gpetiot)
```
Comment (Julow) [inline CHANGES.md:34]:
```suggestion
- Fix ';;' formatting between doc-comments and toplevel directives (#2432, @gpetiot)
```

File: lib/Fmt_ast.ml [hunk 2]
```diff
@@ -4448,7 +4448,11 @@ let fmt_toplevel ?(force_semisemi = false) c ctx itms =
     let last = Option.is_none next in
     let semisemi =
       match (itm, next) with
+      | `Item {pstr_desc= Pstr_attribute _; _}, _ -> false
       | _, Some (`Item {pstr_desc= Pstr_eval _; _}, _) -> true
+      | ( `Item {pstr_desc= Pstr_eval _; _}
+        , Some (`Item {pstr_desc= Pstr_attribute _; _}, _) ) ->
+          true
```
Comment (Julow) [inline lib/Fmt_ast.ml:4455]: There's no `;;` here:

```ocaml
x + 1

let y = 2
```

Is it really needed here ?

```ocaml
x + 1;;

(** foo *)

let y = 2
```

The diffs in test_branch seems to indicate that no.

Comment (gpetiot) [reply lib/Fmt_ast.ml]: I fixed the regressions, let me know what you think.

File: test/passing/tests/docstrings_toplevel_directives.mlt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+(** Header *)
+
+#use "something"
+
+let two = 2
+
+[@@@warning "-labels-omitted"] ;;
```
Comment (Julow) [inline test/passing/tests/docstrings_toplevel_directives.mlt:7]: This `;;` seems unnecessary as well ?

Comment (gpetiot) [reply test/passing/tests/docstrings_toplevel_directives.mlt:7]: Apparently it is, I get a syntax error from the compiler if I remove them.

Comment (Julow) [reply test/passing/tests/docstrings_toplevel_directives.mlt:7]: It is! Sorry about that.

**Review by Julow — APPROVED** [review]
Looks good :) Let's merge this before the release!

**Pull request merged**